daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Upgrading.scala

```diff
@@ -0,0 +1,464 @@
+// Copyright (c) 2024 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.lf
+package validation
+
+import com.daml.lf.data.Ref
+import com.daml.lf.language.Ast
+import scala.util.{Try, Success, Failure}
+import com.daml.lf.validation.AlphaEquiv
+import com.daml.lf.data.ImmArray
+
+case class Upgrading[A](past: A, present: A) {
+  def map[B](f: A => B): Upgrading[B] = Upgrading(f(past), f(present))
+  def fold[B](f: (A, A) => B): B = f(past, present)
+  def zip[B, C](that: Upgrading[B], f: (A, B) => C) =
+    Upgrading(f(this.past, that.past), f(this.present, that.present))
+}
+
+sealed abstract class UpgradeError extends ValidationError {
+  def message: String;
+  def prettyInternal: String = this.message
+  def context: Context = Context.None
+}
+
+final case class CouldNotResolveUpgradedPackageId(packageId: Upgrading[Ref.PackageId])
```

Comment:
Not oblige to follow the following principal, but I like to put such case class inside the companion object of the class they implemented.

Reply:
fixed